src/server/dcp_server/models.py

```diff
@@ -295,3 +297,164 @@ def eval(self, img):
+        self.model_config = model_config["unet"]
+        self.train_config = train_config["unet"]
+        self.eval_config = eval_config["unet"]
+
+        self.in_channels = self.model_config["in_channels"]
+        self.out_channels = self.model_config["out_channels"]
+        self.features = self.model_config["features"]
+        
+        self.encoder = nn.ModuleList()
+        self.decoder = nn.ModuleList()
+
+        self.pool = nn.MaxPool2d(kernel_size=2, stride=2)
+       
+        # Encoder
+        for feature in self.features:
+            self.encoder.append(
+                UNet.DoubleConv(self.in_channels, feature)
+            )
+            self.in_channels = feature
+
+        # Decoder
+        for feature in self.features[::-1]:
+            self.decoder.append(
+                nn.ConvTranspose2d(
+                    feature*2, feature, kernel_size=2, stride=2
+                )
+            )
+            self.decoder.append(
+                UNet.DoubleConv(feature*2, feature)
+            )
+
+        self.bottle_neck = UNet.DoubleConv(self.features[-1], self.features[-1]*2)
+        self.output_conv = nn.Conv2d(self.features[0], self.out_channels, kernel_size=1)
+    
+    def forward(self, x):
+        skip_connections = []
+        for encoder in self.encoder:
+            x = encoder(x)
+            skip_connections.append(x)
+            x = self.pool(x)
+
+        x = self.bottle_neck(x)
+        skip_connections = skip_connections[::-1]
+
+        for i in np.arange(len(self.decoder), step=2):
+            x = self.decoder[i](x)
+            skip_connection = skip_connections[i//2]
+            concatenate_skip = torch.cat((skip_connection, x), dim=1)
+            x = self.decoder[i+1](concatenate_skip)
+
+        return self.output_conv(x)
+
+    def train(self, imgs, masks):
+       
+        lr = self.train_config['lr']
+        epochs = self.train_config['n_epochs']
+        batch_size = self.train_config['batch_size']
+        # optimizer_class = self.train_config['optimizer']
+
+        # Convert input images and labels to tensors
+
+        # normalize images 
+        imgs = [(img-np.min(img))/(np.max(img)-np.min(img)) for img in imgs]
+        # convert to tensor
+        imgs = torch.stack([torch.from_numpy(img.astype(np.float32)) for img in imgs])
+        imgs = imgs.unsqueeze(1) if imgs.ndim == 3 else imgs
+        # Your classification label mask
+        masks = np.array(masks)
+        masks = torch.stack([torch.from_numpy(mask[1]) for mask in masks])
+
+        # Create a training dataset and dataloader
+        train_dataset = TensorDataset(imgs, masks)
+        train_dataloader = DataLoader(train_dataset, batch_size=batch_size)
+
+        loss_fn = nn.CrossEntropyLoss()
+        optimizer = Adam(params=self.parameters(), lr=lr)
+
+        for _ in tqdm(range(epochs), desc="Running UNet training"):
+
+            self.loss = 0
+
+            for _, (imgs, masks) in enumerate(train_dataloader):
```

Comment:
why do enumerate if you are not going to use the index? You can change to: ```for imgs, masks in train_dataloader:```

Reply:
I agree, I was using index to print, but now it's not necessary any more, thanks.